### PR TITLE
Fix S2368 FP: Params keyword on jagged arrays

### LINKED ISSUE #4225: Don't raise S2368 which complains about multidimensional arrays when params keyword is declared on jagged arrays
### Description

The rule S2368 currently fires when the `params` keyword is used on jagged (nested) arrays.

However, at least for two-level jagged arrays declared with `params`, the caller of the function usually just experiences them as a variadic number of distinct arguments of the same type, so the warning should not be triggered in this case.

Additionally, the diagnostics should be fixed to stop calling jagged arrays "multidimensional", as multidimensional arrays (like `byte[,]`) are something different as jagged (nested) arrays (like `byte[][]`). 

### Repro steps

```
namespace SonarLintBugS2368
{
    using System;

    public static class GuidHelper
    {
        public static Guid CreateNameBasedGuid(params byte[][] nameComponents)
        {
            return Guid.Empty;
        }
    }
}
```

### Expected behavior

S2368 does not raise on this method.

### Actual behavior

S2368 is raised on this method, and additionally falsely claims the jagged array to be a multidimensional array.

### Known workarounds

Declare the method as private, and provide N public wrapper methods with 1, 2, 3, ... n parameters of type byte[].

### Related information

* C# Version 8.0 (but works with older versions, too).
* Visual Studio version 2019
* MSBuild / dotnet version: .NET 4.8 / .NET Core 3.1
* SonarScanner for .NET version (if used): SonarScanner for MSBuild 4.9/SonarScanner 4.3.0.2102 (Also reproducible with SonarLint for VS 2019 Version 4.33.1.28913)
* Operating System: Windows 10

### PR DESCRIPTION
**Pull request by csaba-sagi-sonarsource**
Fixes #4225

### CODE REVIEW

**Comment by csaba-sagi-sonarsource** [comment]
Do you agree to update the message of the rule from "Make this method private or simplify its parameters to not use multidimensional arrays." to "Make this method private or simplify its parameters to not use multidimensional or jagged arrays." ?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/PublicMethodWithMultidimensionalArray.vb [hunk 1]
```diff
@@ -38,6 +38,16 @@ Namespace Tests.Diagnostics
...
         End Sub
+
+        Public Sub Method7(ParamArray a As Integer()()) ' Compliant
+        End Sub
+
+        Public Sub Method8(ParamArray a As Integer()()()) ' Noncompliant
+        End Sub
+
+        Sub Method9(ParamArray a As Integer()(,)) ' Compliant
+        End Sub
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/PublicMethodWithMultidimensionalArray.vb:49]: can you add a test with a parameterless declaration? (null param array)

```
    Sub Main
        Console.WriteLine("Hello, world!")
    End Sub
```

File: analyzers/src/SonarAnalyzer.Common/Rules/PublicMethodWithMultidimensionalArrayBase.cs [hunk 2]
```diff
@@ -60,13 +60,15 @@ protected sealed override void Initialize(SonarAnalysisContext context)
                 SyntaxKindsOfInterest.ToArray());
         }
 
-        private static bool MethodHasMultidimensionalArrayParameters(IMethodSymbol methodSymbol)
-        {
-            return methodSymbol.Parameters
-                .Select(param => param.Type as IArrayTypeSymbol)
-                .WhereNotNull()
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/PublicMethodWithMultidimensionalArrayBase.cs:67]: why did you remove the null-check?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/PublicMethodWithMultidimensionalArray.cs [hunk 3]
```diff
@@ -36,6 +36,18 @@ public void MethodX(int[] a)
         public void Method5(int[,] a) //Compliant, implements interface
         {
         }
+
+        void Method6(params int[][,] a) // Compliant
+        {
+        }
+
+        public void Method7(params int[][] a) // Compliant
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/PublicMethodWithMultidimensionalArray.cs:44]: can you also add with
```
public void Method7(int[][] a, params int[][] b) {}
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/PublicMethodWithMultidimensionalArray.cs [hunk 4]
```diff
@@ -36,6 +36,18 @@ public void MethodX(int[] a)
         public void Method5(int[,] a) //Compliant, implements interface
         {
         }
+
+        void Method6(params int[][,] a) // Compliant
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/PublicMethodWithMultidimensionalArray.cs:40]: and also with `public void Method61(int[,][] a) {}`?

**Review by andrei-epure-sonarsource — CHANGES_REQUESTED** [review]
I suggest to add more tests, otherwise looks very nice!

**Review by andrei-epure-sonarsource — APPROVED** [review]
LGTM
